tensorflow/core/kernels/control_flow_ops.cc

```diff
@@ -597,6 +597,15 @@ LoopCondOp::LoopCondOp(OpKernelConstruction* context) : OpKernel(context) {}
 LoopCondOp::~LoopCondOp() = default;
 
 void LoopCondOp::Compute(OpKernelContext* context) {
+  CancellationManager* cm = context->cancellation_manager();
+  bool already_cancelled = cm->IsCancelled();
+
+  if (already_cancelled) {
+    Tensor continue_running(false);
+    context->set_output(0, continue_running);
```

Comment:
After changing it to be fellowing, the `errors::Cancelled("Loop execution was cancelled.")` did not araise as expected. I attach the log of the test case here.

```C++
void LoopCondOp::Compute(OpKernelContext* context) {
  CancellationManager* cm = context->cancellation_manager();
  bool already_cancelled = cm->IsCancelled();
  OP_REQUIRES(context, !already_cancelled,
              errors::Cancelled("Loop execution was cancelled."));

  context->set_output(0, context->input(0));
}
```
```
.......ERROR:tensorflow:Timed out waiting for notification
E..ERROR:tensorflow:Attempting to use uninitialized value Variable
	 [[{{node _retval_Variable_0_0}}]]
..
======================================================================
ERROR: testWhileTimeOutWithControlFlowV2 (__main__.ControlFlowTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/kernel_tests/control_flow_ops_py_test.runfiles/org_tensorflow/tensorflow/python/client/session.py", line 1334, in _do_call
    return fn(*args)
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/kernel_tests/control_flow_ops_py_test.runfiles/org_tensorflow/tensorflow/python/client/session.py", line 1319, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/kernel_tests/control_flow_ops_py_test.runfiles/org_tensorflow/tensorflow/python/client/session.py", line 1407, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.DeadlineExceededError: Timed out waiting for notification

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/kernel_tests/control_flow_ops_py_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 412, in wrapper
    fn(*args, **kwargs)
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/kernel_tests/control_flow_ops_py_test.runfiles/org_tensorflow/tensorflow/python/kernel_tests/control_flow_ops_py_test.py", line 1986, in testWhileTimeOut
    sess.run(r, options=run_options)
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/kernel_tests/control_flow_ops_py_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 1056, in run
    return super(ErrorLoggingSession, self).run(*args, **kwargs)
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/kernel_tests/control_flow_ops_py_test.runfiles/org_tensorflow/tensorflow/python/client/session.py", line 929, in run
    run_metadata_ptr)
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/kernel_tests/control_flow_ops_py_test.runfiles/org_tensorflow/tensorflow/python/client/session.py", line 1152, in _run
    feed_dict_tensor, options, run_metadata)
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/kernel_tests/control_flow_ops_py_test.runfiles/org_tensorflow/tensorflow/python/client/session.py", line 1328, in _do_run
    run_metadata)
  File "/private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/kernel_tests/control_flow_ops_py_test.runfiles/org_tensorflow/tensorflow/python/client/session.py", line 1348, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.DeadlineExceededError: Timed out waiting for notification

----------------------------------------------------------------------
Ran 20 tests in 4.698s

FAILED (errors=1)
```

Reply:
That's working as expected. The external error is first set [here](https://github.com/tensorflow/tensorflow/blob/f25493f9c0ae0dec6db31dfe7fd19ed01b57c23d/tensorflow/core/common_runtime/direct_session.cc#L1699-L1700) before ongoing execution is cancelled, so the `CancelledError` won't bubble up to the client code. However, it's still important to set it to some kind of error, to prevent additional ops from executing (and potentially mutating state based on an incorrect value).